### ISSUE TITLE
Events Stop WasapiOut

### ISSUE THREAD

**Issue by dwebb7** [filoe/cscore#28]
I am in the process of changing a project over from Naudio and and having an issue with events. In my current app I have events fireing to update the meters and waveform drawing. It works fine with Naudio but the event stops the WasapiCapture with CSCore. The example below is a quick project I threw together to show this to you. 

```
    public partial class MainWindow : Window
    {
        private WasapiCapture wasapiCapture;
        private WasapiOut wasapiOut;
        private SoundInSource wasapiCaptureSource;
        private ISampleSource sampleSource;
        private PeakMeter peakMeter;
        private IWaveSource waveSource;
        
        public MainWindow()
        {
            InitializeComponent();
            wasapiCapture = new WasapiCapture();
            wasapiCapture.Initialize();
            wasapiCaptureSource = new SoundInSource(wasapiCapture);
            sampleSource = wasapiCaptureSource.ToSampleSource();
            peakMeter = new PeakMeter(sampleSource);
            peakMeter.Interval = 3000;
            peakMeter.PeakCalculated += peakMeter_PeakCalculated;
            waveSource = peakMeter.ToWaveSource();
            wasapiCapture.Start();
            wasapiOut = new WasapiOut();
            wasapiOut.Initialize(waveSource);
            Thread.Sleep(100);
            wasapiOut.Play();
        }

        void peakMeter_PeakCalculated(object sender, PeakEventArgs e)
        {
            PeakLabel.Content = "test";
        }
    }
```
The above will play the audio from the input for 3seconds until the event fires and then stops because wasapiCapture stops. If I comment out the PeakLabel.Content = "test"; then is works fine. I can even put a Debug.Print there and it will work fine. 
Is this intended behavior?
Thanks

**Comment by dwebb7** [comment]
After looking into this more I have realized that WasapiOut is stopping not WasapiCapture. Is this intended?

**Comment by filoe** [comment]
The problem is that the event does not get fired on the ui thread. Accessing UI Elements from other threads is not allowed. In order to ensure that all components are getting released correctly, the exception gets catched and populated by the Stopped Event of the WasapiOut class. The correct solution for this problem is to use the InvokeAsync Method of any UI Element. It is not recommend to use the sync Invoke Method since it could cause deadlocks or performance critical operations.

**Comment by dwebb7** [comment]
Thanks! That did it. For sake of documentation I updated the code from above with your suggestion using InvokeAsync to change the label to the PeakValue every 100ms. Works great.

```
    public partial class MainWindow : Window
    {
        private WasapiCapture wasapiCapture;
        private WasapiOut wasapiOut;
        private SoundInSource wasapiCaptureSource;
        private ISampleSource sampleSource;
        private PeakMeter peakMeter;
        private IWaveSource waveSource;

        public MainWindow()
        {
            InitializeComponent();
            wasapiCapture = new WasapiCapture();
            wasapiCapture.Initialize();
            wasapiCaptureSource = new SoundInSource(wasapiCapture);
            sampleSource = wasapiCaptureSource.ToSampleSource();
            peakMeter = new PeakMeter(sampleSource);
            peakMeter.Interval = 100;
            peakMeter.PeakCalculated += peakMeter_PeakCalculated;
            waveSource = peakMeter.ToWaveSource();
            wasapiCapture.Start();
            wasapiOut = new WasapiOut();
            wasapiOut.Initialize(waveSource);
            Thread.Sleep(100);
            wasapiOut.Play();
        }

        void peakMeter_PeakCalculated(object sender, PeakEventArgs e)
        {
            Action peakAction = () => lbltest.Content = e.PeakValue.ToString();
            Dispatcher.InvokeAsync(peakAction);
        }
    }
```

**Comment by dwebb7** [comment]
Is this the same issue? I am trying to toggle the reverb mix with a  button click and it stops WasapiOut also. I tried to add a InvokeAsync to the click method and it works sometimes and other times it stops WasapiOut. Any ideas? Here is my code.
```
    public partial class MainWindow : Window
    {
        private WasapiCapture wasapiCapture;
        private WasapiOut wasapiOut;
        private SoundInSource wasapiCaptureSource;
        private readonly DmoWavesReverbEffect reverbEffect;
        private ISampleSource sampleSource;
        private PeakMeter peakMeter;
        private IWaveSource waveSource;

        public MainWindow()
        {
            InitializeComponent();
            wasapiCapture = new WasapiCapture();
            wasapiCapture.Initialize();
            wasapiCaptureSource = new SoundInSource(wasapiCapture);
            sampleSource = wasapiCaptureSource.ToSampleSource();
            peakMeter = new PeakMeter(sampleSource);
            peakMeter.Interval = 100;
            peakMeter.PeakCalculated += peakMeter_PeakCalculated;
            waveSource = peakMeter.ToWaveSource();
            reverbEffect = new DmoWavesReverbEffect(waveSource);
            wasapiCapture.Start();
            wasapiOut = new WasapiOut();
            wasapiOut.Initialize(reverbEffect);
            Thread.Sleep(100);
            wasapiOut.Play();
        }

        void peakMeter_PeakCalculated(object sender, PeakEventArgs e)
        {
            Action peakAction = () => lbltest.Content = e.PeakValue.ToString();
            Dispatcher.InvokeAsync(peakAction);
        }

        private void TestButton_Click(object sender, RoutedEventArgs e)
        {
            Action reverbAction = () =>
            {
                if (reverbEffect.ReverbMix == 0f)
                {
                    reverbEffect.ReverbMix = -96;
                }
                else if (reverbEffect.ReverbMix == -96f)
                {
                    reverbEffect.ReverbMix = 0;
                }
            };
            Dispatcher.InvokeAsync(reverbAction);
        }
    }
```

**Comment by filoe** [comment]
This was a bug which got fixed a few days ago with commit 31bf8cd.
Since the TestButton_Click EventHandler runs on the GUI Thread, it is not necessary to invoke anything.
